### PR TITLE
20 tensor complete documentation for all methods

### PR DESCRIPTION
**Pull request by dmdunla**
Updating docstrings, doctests.

----
:books: Documentation preview :books:: https://pyttb--216.org.readthedocs.build/en/216/

### CODE REVIEW

**Comment by ntjohnson1** [comment]
Is this waiting on review from me or still WIP? I assume the sptensor bug came up when looking over our documentation stuff

**Comment by dmdunla** [comment]
Thanks, still working ... I'll explicitly request a review when ready.

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:62]:
```suggestion
        Creates a :class:`pyttb.tensor` from a :class:`numpy.ndarray`
```

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:145]: The argument in tensor is `function_handle` same with `sptensor`. Unless you find this terribly verbose we should update ktensor to match. Also `required` is unnecessary in the description since there is no default provided.

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:149]:
```suggestion
            `numpy.zeros`, `numpy.ones`.
```

Same typo exists in ktensor.

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:355]:
```suggestion
        Convert `:class:pyttb.tensor` to an `:class:numpy.ndarray` of doubles.
```

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:272]: Is there a particular reason to prefer `i1, i2` over `i,j`? I don't have a strong rationale but `i`, `j` looks nicer to me.

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:421]:
```suggestion
        Contruct a :class:`pyttb.sptensor` from `:class:pyttb.tensor`
```

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:459]:
```suggestion
        Efficient inner product between a tensor and other `pyttb` tensors
```

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:2374]:
```suggestion
    --------
```

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:85]: You need a newline between the description and code for it to render correctly https://pyttb--216.org.readthedocs.build/en/216/tensor.html#pyttb.tensor

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:377]: I think for the multiline returns it needs to be indendented to render correctly https://pyttb--216.org.readthedocs.build/en/216/tensor.html#pyttb.tensor.exp

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:395]: Same indentations rendering note

File: pyttb/tensor.py
Comment (ntjohnson1) [inline pyttb/tensor.py:2245]: An alternative is to use the `...` doctest description to avoid comparing the full float. I think we use that a variety of other places, for example nvecs

File: docs/source/sptensor.rst
Comment (ntjohnson1) [inline docs/source/sptensor.rst:4]: Switching automodule to autoclass drops the auto documentation for the `tt_to_sparse_matrix` and `tt_from_sparse_matrix`. If these are going away/mostly internal then that's ok. Mostly just an fyi

File: pyttb/tensor.py
Comment (ntjohnson1) [reply pyttb/tensor.py:145]: Hmm I see later on we used `fun` instead of the verbose `function_handle` in tensor. We should probably unify globally or file an issue to do so.

**Review by ntjohnson1 — COMMENTED** [review]
Looks good to me. A variety of small fixes/notes.

One global style note. I think it is nice to have full print outs of tensor contents to be explicit, however it does add to maintenance efforts if we ever change the format of the display. Luckily doctest will catch it pretty immediately. I guess since you are the one who just did the doc updates you are in the best position to state if that's a worthwhile tradeoff.

**Comment by dmdunla** [comment]
> Looks good to me. A variety of small fixes/notes.
> 
> One global style note. I think it is nice to have full print outs of tensor contents to be explicit, however it does add to maintenance efforts if we ever change the format of the display. Luckily doctest will catch it pretty immediately. I guess since you are the one who just did the doc updates you are in the best position to state if that's a worthwhile tradeoff.

I went back and forth on this. I think this provides users with examples of what they should see. Since we clear the tutorial notebooks, these docs provide a way to see what pyttb does before you start running it (or while you are trying to determine if it is doing what you want).

File: docs/source/sptensor.rst
Comment (dmdunla) [reply docs/source/sptensor.rst:4]: These will go away once sptenmat is in place. I tried to align with ktensor, which provided cleaner docs, IMHO.

File: pyttb/tensor.py
Comment (dmdunla) [reply pyttb/tensor.py:2245]: That did not work in all places, but I can try again. I am using Github codespaces (while traveling) and there seem to be some differences from my typically VS code setup.

File: pyttb/tensor.py
Comment (dmdunla) [reply pyttb/tensor.py:272]: I was trying to be explicit with the ordering. This follows papers that that move away from i,j,k that is used for 3-way tensors.

File: docs/source/sptensor.rst
Comment (dmdunla) [reply docs/source/sptensor.rst]: I went back to automodule to include the methods in each of the class files. The class constructor is a bit confusing now, though, at least to me. It states `pyttb.tensor.tensor(...)`, instead of `pyttb.tensor(...)` as it does when stating autoclass.

@ntjohnson1 any thoughts on this?

File: pyttb/tensor.py
Comment (ntjohnson1) [reply pyttb/tensor.py]: This wasn't actually resolved but I just pushed a commit that should have fixed the formatting.

File: docs/source/sptensor.rst
Comment (ntjohnson1) [reply docs/source/sptensor.rst]: No strong thoughts. If you think the autoclass is cleaner we can probably stick to that for now. This relates to discussions we've had in the past about slightly ambiguous code layout because of our flat structure mapping to the MATLAB version. Probably doesn't need to be resolved for this PR though. We could also pull out some of the other methods from the class files and think about a better place for them so we could use autoclass AND capture the documentation for these methods.

Some more code organization thoughts:
* It's a little confusing we use tensor generically for all multi-dim data but also tensor specifically for dense tensors
* It adds to the ambiguity that we use lowercase class names (to match MATLAB)
* We add convenience imports to flatten the structure, but the docs probably should refer to the actual source location (I am not sure if there is an easy way to convince sphinx otherwise)

Related to the first two, in my repo where I started to play with a rust port (that I haven't touched in months). I was playing with a structure that looks like
```
ttb/
   tensors/
      Dense
      Kruskal
      Sparse
      Tucker
```

**Review by ntjohnson1 — APPROVED** [review]
I think for the intention of adding full docs coverage to tensor this PR looks good. There are a few open points, but those can probably get moved to a follow on ticket/discussion.

**Pull request merged**
